### PR TITLE
Extract and create new PWM library

### PR DESCRIPTION
**Pull request by pkendall64**
This is another PR that is in preparation for S3 support.

The S3 only has 8 `LEDC` channels, unlike the ESP32 which has 16. Both also have an `MCPWM` module (Motor Control PWM) with 12 channels!

This PR extracts the PWM code out of the SERVO library into a seperate PWM library (with API documentation) which is used by all devices which need a PWM output signal. Also implemented in the library is ESP32 support using the `MCPWM` module and the `LEDC` module, providing the capability for up to 20 PWM channels!

| ESP8266 | ESP32 (and S3) | 
| -- | -- |
| Up to 13 channels | Up to 20 channels |
| Based on waveform_8266, controlled via a hardware timer | Channels are allocated based on update frequency, and allocated to the MCPWM first, then 8 LEDC channels |

The Servo code has been refactored to make use of the new PWM library and external dependencies on the following libraries has been removed.
- roboticsbrno/ServoESP32 @ 1.0.3
- dlloydev/ESP32 ESP32S2 AnalogWrite @ 3.0.3

This refactored code has been tested on an ESP32 based Super-P with 12 servos, an ESP32-S3 test unit with 16 servos and an ESP8285 based PWM receiver with 5 servos.

### CODE REVIEW

**Comment by hydra** [comment]
I'll be testing this on my new S3 based design once it's assembled, hopefully within in a week or two.

File: src/lib/PWM/PWM.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     * @brief Allocate a pin/frequency to a PWM channel
+     *
+     * @param pin the GPIO pin to output the PWM signal on
+     * @param frequency the output frequency of the PWM signal
+     * @return the allocated PWM channel or -1 if a channel cannot be assigned
+     */
+    pwm_channel_t allocate(uint8_t pin, uint32_t frequency);
+
+    /**
+     * @brief release a channel back to the allocater
```
Comment (wvarty) [inline src/lib/PWM/PWM.h:18]: allocator

File: src/lib/PWM/PWM_ESP32.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+pwm_channel_t PWMController::allocate(uint8_t pin, uint32_t frequency)
+{
+    // 1. see if we can allocate a MCPWM channel at this frequency
+    int channel = -1;
+    // 1a. see if theres a MCPWM already using this frequency we can piggy-back on
+    for (int i=0 ; i<MCPWM_CHANNELS ; i++)
+    {
+        if (mcpwm_frequencies[i] == frequency)
+        {
+            if (i%2 == 0 && mcpwm_frequencies[i+1]==0)
```
Comment (wvarty) [inline src/lib/PWM/PWM_ESP32.cpp:97]: The inconsistency WRT whitespace between the `==` is mildly triggering 😅

Comment (pkendall64) [reply src/lib/PWM/PWM_ESP32.cpp]: I ran the code formatter so it's fixed all the similar problems too.

**Review by JyeSmith — APPROVED** [review]

**Review by wvarty — APPROVED** [review]

**Pull request merged**